### ISSUE TITLE
SSH connection closed after closing the webpage?

### ISSUE THREAD

**Issue by P4sca1** [phpseclib/phpseclib#1194]
I have installed phpseclib on my Apache 2 webserver with PHP 7.1.
I am using it to tail -f a log file and output the contents directly to the browser. Therefore I'm setting the php time limit to 0 and the ssh timeout also to 0.
My frontend then makes a request with oboe.js to the php page, keeps the connection open and reads the output of the php file to render a live console.

I fear that I have hundreds of ssh connections opened at the end, so is the ssh connection closed after the user closes the browser tab? Also is there a good way to also stop tail -f when the ssh connection is closed with phpseclib? I have read that connecting with the -t flag should help somehow.

**Comment by terrafrost** [comment]
phpseclib will close the SSH connection if the destructor is called. If the tab were closed the PHP process would probably end, which would result in the destructor being called, so yah, the ssh connection should close and the command should end.

I guess ultimately it depends on how you have PHP configured. You could do `ps aux | grep php` I guess if you had doubts that PHP processes weren't ending but idk that that'd always work (eg. mod_php vs CGI, idk).

**Comment by P4sca1** [comment]
Thank you very much for your answer.
I checked my /var/log/auth.log and this is what I noticed.

``` without timeout, closing the page after a few seconds manually
Oct  8 17:18:06 ds1 sshd[24717]: Accepted password for master from  {ip-address} port 54477 ssh2
Oct  8 17:18:06 ds1 sshd[24717]: pam_unix(sshd:session): session opened for user master by (uid=0)
Oct  8 17:18:06 ds1 sudo:   master : TTY=unknown ; PWD=/home/master ; USER=egm-dev-game-1-1 ; COMMAND=/home/master/scripts/useraction.sh gameserver_tail_screenlog gmod
Oct  8 17:18:06 ds1 sudo: pam_unix(sudo:session): session opened for user egm-dev-game-1-1 by (uid=0)
```

``` timeout 5 seconds
Oct  8 17:19:43 ds1 sshd[25012]: Accepted password for master from  {ip-address} port 54859 ssh2
Oct  8 17:19:43 ds1 sshd[25012]: pam_unix(sshd:session): session opened for user master by (uid=0)
Oct  8 17:19:43 ds1 sudo:   master : TTY=unknown ; PWD=/home/master ; USER=egm-dev-game-1-1 ; COMMAND=/home/master/scripts/useraction.sh gameserver_tail_screenlog gmod
Oct  8 17:19:43 ds1 sudo: pam_unix(sudo:session): session opened for user egm-dev-game-1-1 by (uid=0)
Oct  8 17:19:48 ds1 sshd[25018]: Received disconnect from  {ip-address}: 11:
Oct  8 17:19:48 ds1 sshd[25012]: pam_unix(sshd:session): session closed for user master
```

From what I can see, the ssh connection without a timeout is never closed. Could it be because of set_time_limit(0)?

If so, do you know any other way to keep a ssh connection open while the browser tab is open?

**Comment by P4sca1** [comment]
Oh, I noticed another thing. The connection is indeed closed, exactly 2 minutes after it has been created!
I tried it another time, and it was exactly 2 minutes again.

**Comment by P4sca1** [comment]
Okay I found the reason.
Referring to the php docs (http://php.net/manual/de/features.connection-handling.php), php is calling the shutdown function and thus disconnecting the ssh session, when noticing, that the user has closed the page. PHP checks that everytime, the script is giving some output. In my case I am using tail and if there is no output, the connection will persist, even if the user has closed the page. I manually inserted something to the file tail was referencing it, and the ssh connection immediately got closed.

So the solution to my problem would be to either hope for tail to having output or checking the status in an infinite loop with connection_aborted() (http://php.net/manual/de/function.connection-aborted.php) and when it is aborted call exit(), which then calls the shutdown function and disconnects ssh.

So phpseclib is working just as expected and is closing the connection right.
Anyways, thanks for your answer! 👍
